accepted/2020/mono-convergence/platform-specific-httpclient.md

```diff
@@ -0,0 +1,452 @@
+
+**Conclusion**
+
+The big advantage of this proposal is that the default HttpClientHandler would be have the most optimal implementation on every platform (performance and size wise). The availability of only platform specific properties on HttpClientHandler has its pros & cons as well.
+
+### Proposal #5: All HttpHandlers are fully implemented inbox
+
+Implement all platform-specific handlers fully inside dotnet/runtime using platform-specific interop libraries. We would extend low-level interop API which and use it to implement the managed API for HttpClientHandler and other types which are exposed from it.
+
+There is already a foundation for CF interop at [https://github.com/dotnet/runtime/tree/master/src/libraries/Common/src/Interop/OSX](https://github.com/dotnet/runtime/tree/master/src/libraries/Common/src/Interop/OSX). This will need to be extended to include Objective-C bridge for necessary APIs.
+
+We could also expose this Interop layer to Xamarin internally to avoid code duplication as they use the same platform API.
+
+**Disadvantages**
+
+* Implementation of HttpClientHandler might require implementing more dependent type (e.g. Android.Net.SSLSocketFactory)
+* Problems with handling platform-specific exceptions
+* Not all decorations used for Xamarin bindings will be available (e.g. [iOS (7,0)])
+* Starting from scratch for Android
+
+**Conclusion**
+
+No build magic and more shared code prospect, as well as the ability to test HttpClient changes easily across all platforms, are very appealing.
+
+## The plan for .NET 5
+
+The decision is to use different approaches for each platform for .NET 5. We might consider unifying the approaches eventually but not in .NET 5 time-frame.
+
+A new feature like setting will be introduced which will allow controlling the implicit handler selection. The internal property will be available inside System.Net.Http namespace and mapped to msbuild property or AppContext setting.
+
+```c#
+namespace System.Net.Http
+​​{
+​​  public partial class HttpClientHandler
+​​  {
+​​    internal static HandlerType DefaultHandlerType { get; } = HttpClientHandlerType.PlatformNative;
+​​  }
+}
+```
+
+The implementation has to use the property in the way to allow the ILLinker to remove the unnecessary dependency of unused handler when the type is not used elsewhere.
+
+### Xamarin.iOS/Xamarin.tvOS
+Follow proposal #2b and hook up existing `NSUrlSessionHandler` using `Activator.CreateInstance ("Foundation.NSUrlSessionHandler, Xamarin.iOS/Xamarin.tvOS")​`. The actual implementation will remain inside xamarin/xamarin-macios​ in repo.
+
+The plan ignores CFNetworkHandler and other legacy handlers. They will stay in `xamarin/xamarin-macios` and will be available only when constructed manually. The implementation logic inside HttpClientHandler will rely on DefaultHandlerType selection to set the implicit handler behaviour between NSUrlSessionHandler and SocketHttpHandler.
```

Comment:
`CFNetworkHandler` is the only native option for iOS7.x

Reply:
I don't think this matters for netcoreapp side